horovod/run/elastic/discovery.py

```diff
@@ -0,0 +1,166 @@
+# Copyright 2020 Uber Technologies, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+from __future__ import absolute_import
+
+import logging
+import threading
+
+from collections import defaultdict
+
+import six
+
+from horovod.run.common.util import safe_shell_exec
+
+
+class HostState(object):
+    def __init__(self):
+        self._event = threading.Event()
+
+        # TODO(travis): blacklisted hosts should have a timeout period that increases with each failure
+        self._blacklisted = False
+
+    def get_event(self):
+        if self._event.is_set():
+            event = threading.Event()
+            self._event = event
+        return self._event
+
+    def set_event(self):
+        self._event.set()
+
+    def blacklist(self):
+        self._blacklisted = True
+        self.set_event()
+
+    def is_blacklisted(self):
+        return self._blacklisted
+
+
+class DiscoveredHosts(object):
+    def __init__(self, host_slots, host_assignment_order):
+        self._host_slots = host_slots
+        self._host_assignment_order = host_assignment_order
+
+    @property
+    def host_slots(self):
+        return self._host_slots
+
+    @property
+    def available_hosts(self):
+        return set(self._host_assignment_order)
+
+    @property
+    def host_assignment_order(self):
+        return self._host_assignment_order
+
+    def get_slots(self, host):
+        return self._host_slots.get(host, 0)
+
+    def count_available_slots(self):
+        # Use the host_assignment_order as it does not contain blacklisted hosts
+        return sum([self.get_slots(host) for host in self._host_assignment_order])
+
+    def update(self, hosts_state):
+        self._host_assignment_order = [host for host in self._host_assignment_order
+                                       if not hosts_state[host].is_blacklisted()]
+        return self
+
+
+class HostManager(object):
```

Comment:
with a mocked discovery, the whole class should be nicely unit-testable (changing set of discovered hosts, changing order of hosts, blacklisting hosts)

Reply:
Good idea.  Added tests.